components/mediators/cache/org.wso2.carbon.mediator.cache/src/main/java/org/wso2/carbon/mediator/cache/util/HttpCachingFilter.java

```diff
@@ -0,0 +1,196 @@
+/*
+ *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+package org.wso2.carbon.mediator.cache.util;
+
+import com.google.common.net.HttpHeaders;
+import org.apache.axis2.Constants;
+import org.apache.commons.collections.map.MultiValueMap;
+import org.apache.commons.lang.StringUtils;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.transport.nhttp.NhttpConstants;
+import org.wso2.carbon.mediator.cache.CachableResponse;
+import org.wso2.carbon.mediator.cache.CachingConstants;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is used to apply the filter.
+ */
+public class HttpCachingFilter {
+
+    private HttpCachingFilter() {
+    }
+
+    /**
+     * If the cached response is expired, response needs to be fetched again. And if the cached response have the
+     * no-cache header, existing cached response needs to be validated with the backend by sending a request with
+     * If-None-Match header with ETag value. This method returns whether a request needs to be sent to the backend
+     * or not.
+     *
+     * @param cachedResponse The cached response.
+     * @param synCtx         The message context.
+     * @return True if response need to validated with backend.
+     */
+    public static boolean isValidCacheEntry(CachableResponse cachedResponse, MessageContext synCtx) {
+        Map<String, Object> httpHeaders = cachedResponse.getHeaderProperties();
```

Comment:
Recently we have changed the type of headerProperties to ConcurrentHashMap in https://github.com/wso2/carbon-mediation/commit/db65eb3e00d1b83c4cb1146943697fe288530cd3#diff-b4c37deafefe63bf3b584a03db8aa63eR61

Better to test for https://github.com/wso2/product-ei/issues/1853